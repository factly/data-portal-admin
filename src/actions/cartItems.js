import axios from '../utils/axios';
import {
  ADD_CARTITEMS,
  SET_CARTITEM_LOADING,
  SET_CARTITEM_REQUEST,
  SET_CARTITEM_IDS,
  CARTITEM_API,
} from '../constants/cartItems';
import { addProducts } from './products';
import { getIds, getValues, deleteKeys, buildObjectOfItems } from '../utils/objects';

export const loadCartItems = (id, page = 1, limit = 5) => {
  return async (dispatch, getState) => {
    const {
      cartItems: { req },
    } = getState();

    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
      }
    }

    if (ids) {
      dispatch(setCartItemIds(ids));
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${CARTITEM_API(id)}?page=${page}&limit=${limit}`,
      method: 'get',
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { id, page: page, limit: limit, ids: currentPageIds };
    dispatch(setCartItemRequest(currentReq, total));
    dispatch(addCartItems(nodes));
    dispatch(setCartItemIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

const setLoading = (loading) => {
  return {
    type: SET_CARTITEM_LOADING,
    payload: { loading },
  };
};

export const addCartItems = (cartItems) => (dispatch) => {
  const products = getValues(cartItems, 'product');
  dispatch(addProducts(products));

  dispatch({
    type: ADD_CARTITEMS,
    payload: {
      cartItems: buildObjectOfItems(deleteKeys(cartItems, ['product'])),
    },
  });
};

const setCartItemRequest = (req, total) => {
  return {
    type: SET_CARTITEM_REQUEST,
    payload: { req, total },
  };
};

const setCartItemIds = (ids) => {
  return {
    type: SET_CARTITEM_IDS,
    payload: { ids },
  };
};
