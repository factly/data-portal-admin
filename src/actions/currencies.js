import axios from '../utils/axios';
import {
  ADD_CURRENCY,
  ADD_CURRENCIES,
  SET_CURRENCY_LOADING,
  SET_CURRENCY_REQUEST,
  SET_CURRENCY_IDS,
  RESET_CURRENCY,
  CURRENCY_API,
} from '../constants/currencies';
import { getIds, buildObjectOfItems } from '../utils/objects';

export const loadCurrencies = (page = 1, limit = 5) => {
  return async (dispatch, getState) => {
    const {
      currencies: { req },
    } = getState();

    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
      }
    }

    if (ids) {
      dispatch(setCurrencyIds(ids));
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${CURRENCY_API}?page=${page}&limit=${limit}`,
      method: 'get',
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { page: page, limit: limit, ids: currentPageIds };
    dispatch(setCurrencyRequest(currentReq, total));
    dispatch(addCurrencies(nodes));
    dispatch(setCurrencyIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

export const createCurrency = (data) => {
  return async (dispatch, getState) => {
    dispatch(setLoading(true));

    await axios({
      url: CURRENCY_API,
      method: 'post',
      data: data,
    });

    dispatch(resetCurrency());
    dispatch(setLoading(false));
  };
};

export const updateCurrency = (id, data) => {
  return async (dispatch, getState) => {
    let url = `${CURRENCY_API}/${id}`;

    dispatch(setLoading(true));

    const response = await axios({
      url: url,
      method: 'put',
      data: data,
    });

    dispatch(addCurrency(response.data));
    dispatch(setLoading(false));
  };
};

export const deleteCurrency = (id) => {
  return async (dispatch, getState) => {
    let url = `${CURRENCY_API}/${id}`;

    dispatch(setLoading(true));

    await axios({
      url: url,
      method: 'delete',
    });

    dispatch(resetCurrency());
    dispatch(setLoading(false));
  };
};

export const getCurrency = (id) => {
  return async (dispatch, getState) => {
    const {
      currencies: { items },
    } = getState();

    if (items[id]) {
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${CURRENCY_API}/${id}`,
      method: 'get',
    });
    dispatch(addCurrency(response.data));

    dispatch(setLoading(false));
  };
};

const setLoading = (loading) => {
  return {
    type: SET_CURRENCY_LOADING,
    payload: { loading },
  };
};

const addCurrency = (currency) => {
  return {
    type: ADD_CURRENCY,
    payload: { currency },
  };
};

export const addCurrencies = (currencies) => {
  return {
    type: ADD_CURRENCIES,
    payload: {
      currencies: buildObjectOfItems(currencies),
    },
  };
};

const setCurrencyRequest = (req, total) => {
  return {
    type: SET_CURRENCY_REQUEST,
    payload: { req, total },
  };
};

const setCurrencyIds = (ids) => {
  return {
    type: SET_CURRENCY_IDS,
    payload: { ids },
  };
};

const resetCurrency = () => {
  return {
    type: RESET_CURRENCY,
  };
};
