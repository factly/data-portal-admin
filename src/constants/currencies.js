export const ADD_CURRENCY = 'ADD_CURRENCY';
export const ADD_CURRENCIES = 'ADD_CURRENCIES';
export const SET_CURRENCY_LOADING = 'SET_CURRENCY_LOADING';
export const SET_CURRENCY_REQUEST = 'SET_CURRENCY_REQUEST';
export const SET_CURRENCY_IDS = 'SET_CURRENCY_IDS';
export const RESET_CURRENCY = 'RESET_CURRENCY';

export const CURRENCY_API = `/currencies`;
