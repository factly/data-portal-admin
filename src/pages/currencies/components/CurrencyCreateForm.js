import React from 'react';
import { useHistory } from 'react-router-dom';
import { Form, Input, Button, notification } from 'antd';

const formItemLayout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 14 },
};

const CurrencyCreateForm = ({ onSubmit, data = {} }) => {
  const history = useHistory();

  const onFinish = (values) => {
    onSubmit(values)
      .then(() => {
        notification.success({
          message: 'Success',
          description: 'Currency succesfully added',
        });
        history.push('/currencies');
      })
      .catch(() => {
        notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
  };

  return (
    <Form name="currencies_create" initialValues={data} {...formItemLayout} onFinish={onFinish}>
      <Form.Item
        label="Name"
        name="name"
        rules={[
          {
            required: true,
            message: 'Please enter name!',
          },
        ]}
      >
        <Input placeholder="Ex. Indian Rupee" />
      </Form.Item>

      <Form.Item
        label="ISO Code"
        name="iso_code"
        rules={[
          {
            required: true,
            message: 'Please enter iso code!',
          },
        ]}
      >
        <Input placeholder="Ex. INR" />
      </Form.Item>

      <Form.Item wrapperCol={{ span: 12, offset: 6 }}>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

export default CurrencyCreateForm;
