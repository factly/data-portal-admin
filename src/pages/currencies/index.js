import React from 'react';
import CurrenciesList from './components/CurrenciesList';
import { Button } from 'antd';
import { Link } from 'react-router-dom';

const Currencies = () => {
  return (
    <>
      <Link to={'/currencies/create'}>
        <Button type="primary" style={{ marginBottom: 16 }}>
          Add Currency
        </Button>
      </Link>
      <CurrenciesList />
    </>
  );
};

export default Currencies;
