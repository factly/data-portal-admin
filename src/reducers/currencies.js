import produce from 'immer';
import {
  ADD_CURRENCY,
  ADD_CURRENCIES,
  SET_CURRENCY_LOADING,
  SET_CURRENCY_REQUEST,
  SET_CURRENCY_IDS,
  RESET_CURRENCY,
} from '../constants/currencies';

const initialState = {
  loading: false,
  ids: [],
  req: [],
  items: {},
  total: 0,
};

const currenciesReducer = produce((draft, action = {}) => {
  switch (action.type) {
    case SET_CURRENCY_LOADING:
      draft.loading = action.payload.loading;
      return;
    case ADD_CURRENCY: {
      const { currency } = action.payload;
      draft.items[currency.id] = currency;
      return;
    }
    case ADD_CURRENCIES: {
      const { currencies } = action.payload;
      Object.assign(draft.items, currencies);
      return;
    }
    case SET_CURRENCY_IDS:
      draft.ids = action.payload.ids;
      return;
    case SET_CURRENCY_REQUEST: {
      const { req, total } = action.payload;
      draft.req.push(req);
      draft.total = total;
      return;
    }
    case RESET_CURRENCY:
      return initialState;
  }
}, initialState);

export default currenciesReducer;
