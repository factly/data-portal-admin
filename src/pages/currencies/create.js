import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Form, Input, Button, notification } from 'antd';

import { createCurrency } from '../../actions/currencies';

const formItemLayout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 14 },
};

const CurrencyCreate = (props) => {
  const { create } = props;

  const onFinish = (values) => {
    create(values)
      .then(() => {
        notification.success({
          message: 'Success',
          description: 'Currency succesfully added',
        });
        props.history.push('/currencies');
      })
      .catch(() => {
        notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
  };

  return (
    <Form name="currencies_create" {...formItemLayout} onFinish={onFinish}>
      <Form.Item
        label="Name"
        name="name"
        rules={[
          {
            required: true,
            message: 'Please enter name!',
          },
        ]}
      >
        <Input placeholder="Ex. Indian Rupee" />
      </Form.Item>

      <Form.Item
        label="ISO Code"
        name="iso_code"
        rules={[
          {
            required: true,
            message: 'Please enter iso code!',
          },
        ]}
      >
        <Input placeholder="Ex. INR" />
      </Form.Item>

      <Form.Item wrapperCol={{ span: 12, offset: 6 }}>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

CurrencyCreate.propTypes = {
  create: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  create: (values) => dispatch(createCurrency(values)),
});

export default connect(null, mapDispatchToProps)(CurrencyCreate);
