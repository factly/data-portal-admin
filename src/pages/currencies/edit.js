import React from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';

import CurrencyCreateForm from './components/CurrencyCreateForm';
import { getCurrency, updateCurrency } from '../../actions/currencies';

const EditCurrency = () => {
  const { id } = useParams();
  const history = useHistory();
  const dispatch = useDispatch();
  const { currency } = useSelector(({ currencies }) => currencies);

  React.useEffect(() => {
    dispatch(getCurrency(id));
  }, []);

  if (!currency.id) return <Skeleton />;

  const onUpdate = async (values) => {
    await dispatch(updateCurrency(id, values));
    history.push('/currencies');
  };

  return <CurrencyCreateForm data={currency} onSubmit={onUpdate} />;
};

export default EditCurrency;
