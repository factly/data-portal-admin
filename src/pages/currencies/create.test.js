import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';
import { act } from '@testing-library/react';

import '../../matchMedia.mock';
import CreateCurrency from './create';
import * as actions from '../../actions/currencies';
import CurrencyCreateForm from './components/CurrencyCreateForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

jest.mock('../../actions/currencies', () => ({
  createCurrency: jest.fn(),
}));

describe('Currencies create component', () => {
  let store;
  let mockedDispatch;

  store = mockStore({
    currencies: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
  });
  store.dispatch = jest.fn(() => ({}));
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);
  describe('snapshot testing', () => {
    it('should render the component', () => {
      const component = shallow(<CreateCurrency />);
      expect(component).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should call createCurrency', (done) => {
      actions.createCurrency.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <CreateCurrency />
          </Provider>,
        );
      });
      wrapper.find(CurrencyCreateForm).props().onSubmit({ test: 'test' });
      setTimeout(() => {
        expect(actions.createCurrency).toHaveBeenCalledWith({ test: 'test' });
        expect(push).toHaveBeenCalledWith('/currencies');
        done();
      }, 0);
    });
  });
});
