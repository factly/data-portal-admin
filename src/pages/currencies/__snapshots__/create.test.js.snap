// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currencies create component snapshot testing should render the component 1`] = `
<CurrencyCreateForm
  onSubmit={[Function]}
/>
`;
